ftw/testbrowser/drivers/webtest.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import absolute_import
+from copy import deepcopy
+from ftw.testbrowser.drivers.utils import isolated
+from ftw.testbrowser.drivers.utils import remembering_for_reload
+from ftw.testbrowser.exceptions import BlankPage
+from ftw.testbrowser.exceptions import RedirectLoopException
+from ftw.testbrowser.interfaces import IDriver
+from ftw.testbrowser.utils import copy_docs_from_interface
+from webtest import TestApp
+from zope.interface import implementer
+from ZPublisher.WSGIPublisher import publish_module
+from ZPublisher.WSGIPublisher import set_default_debug_exceptions
+import six
+
+
+@isolated
+def wsgi_app(environ, start_response):
+    return publish_module(environ, start_response)
+
+
+@copy_docs_from_interface
+@implementer(IDriver)
+class WebtestDriver(object):
+
+    LIBRARY_NAME = 'webtest library'
+    WEBDAV_SUPPORT = True
+
+    def __init__(self, browser):
+        self.browser = browser
+        self.app = TestApp(wsgi_app)
+        self.max_redirects = 5
+        self.reset()
+
+    def reset(self):
+        self.response = None
+        self.current_url = None
+        self.headers = {}
+        self.no_redirects = 0
+        self.previous_make_request = None
+        self.app.reset()
+
+    @remembering_for_reload
+    def make_request(self, method, url, data=None, headers=None,
+                     referer_url=None):
+        self.current_url = url
+        headers = headers or {}
+
+        if referer_url and referer_url.strip():
+            headers['REFERER'] = referer_url
+
+        headers.update(self.headers)
+
+        if self.browser.exception_bubbling:
+            set_default_debug_exceptions(True)
+        else:
+            set_default_debug_exceptions(False)
+
+        if method.upper() == 'GET':
+            self.response = self.app.get(
+                url, params=data, headers=headers, expect_errors=True)
+        elif method.upper() in ['POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS',
+                                'HEAD']:
+            impl = getattr(self.app, method.lower())
+            data = self._encode_data(data)
+            self.response = impl(
+                url, params=data or '', headers=headers, expect_errors=True)
+        else:
+            self.response = self.app.request(
+                url, expect_errors=True, method=method)
+
+        if (self.browser.allow_redirects
+                and self.response.status_code // 100 == 3):
+            return self.follow_redirect(
+                method, url, data=data, headers=headers)
+        return (
+            self.response.status_code,
+            self.response.status[4:],
+            self.response.body,
+        )
+
+    def follow_redirect(self, method, url, data=None, headers=None,
+                        referer_url=None):
+        if self.no_redirects > self.max_redirects:
+            raise RedirectLoopException(url)
+        self.no_redirects += 1
```

Comment:
The variable name confused me because I read "no redirects" literally, not "number of redirects".
Maybe we should rename that to `num_redirects`?

Reply:
renamed